Tribler/community/multichain/community.py

```diff
@@ -0,0 +1,251 @@
+                     "a81fdcb2db302ac15905d6e75f96b39ccdaf068bdbbda81a6356f53f7ce4e"
+        master_key_hex = master_key.decode("HEX")
+        master = dispersy.get_member(public_key=master_key_hex)
+        return [master]
+
+    def initiate_meta_messages(self):
+        return super(MultiChainCommunity, self).initiate_meta_messages() + [
+            Message(self, SIGNATURE,
+                    DoubleMemberAuthentication(
+                        allow_signature_func=self.allow_signature_request, split_payload_func=split_function),
+                    PublicResolution(),
+                    DirectDistribution(),
+                    CandidateDestination(),
+                    SignaturePayload(),
+                    self._generic_timeline_check,
+                    self.received_signature_response)]
+
+    def initiate_conversions(self):
+        return [DefaultConversion(self), MultiChainConversion(self)]
+
+    def publish_signature_request_message(self, candidate):
+        """
+        Creates and sends out a signed signature_request message.
+        """
+        self._logger.info("Sending signature request.")
+        # The lock is lifted after the timeout or a valid signature response is received.
+        self._logger.info("Get Lock: send signature request: %s" % self.chain_lock.locked())
+        self.chain_lock.acquire()
+        self._logger.info("Acquired Lock: send signature request.")
+
+        message = self.create_signature_request_message(candidate)
+        self.create_signature_request(candidate, message, self.allow_signature_response, timeout=5.0)
+
+    def create_signature_request_message(self, candidate):
+        """
+        Create a signature request message using the current time stamp.
+        :return: Signature_request message ready for distribution.
+        """
+        # Instantiate the data
+        up = 1
+        down = 2
+        total_up = 3
+        total_down = 4
+        # Instantiate the personal information
+        sequence_number_requester = self._get_next_sequence_number()
+        previous_hash_requester = self._get_latest_hash()
+
+        payload = (up, down, total_up, total_down, sequence_number_requester, previous_hash_requester)
+        meta = self.get_meta_message(SIGNATURE)
+
+        message = meta.impl(authentication=([self.my_member, candidate.get_member()],),
+                            distribution=(self.claim_global_time(),),
+                            payload=payload)
+        return message
+
+    def allow_signature_request(self, message):
+        """
+        We've received a signature request message, we must either:
+            a. append to this message our data (Afterwards we sign the message.).
+            b. None (if we want to drop this message)
+        """
+        self._logger.info("Received signature request.")
+        self._logger.info("Chain Lock: process request: %s" % self.chain_lock.locked())
+        if self.chain_lock.acquire(False):
```

Comment:
There is a test with data loss in the tests. There is a test with a client ignoring the signature request. This is the same situation as a attacker not replying, signature response being lost, or a signature request being lost.

Locking is still needed because of incoming requests for A. Next to that, the code should not be dependent on timing issues to work correctly. 

An inconsistency is allowed in B's chain. 

Alice signing last creates the same problem, but then for Bob having to wait upon Alice's response.